sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala

```diff
@@ -497,12 +497,10 @@ final class DataFrameNaFunctions private[sql](df: DataFrame) {
           throw new IllegalArgumentException(s"$targetType is not matched at fillValue")
       }
       // Only fill if the column is part of the cols list.
-      if (typeMatches && cols.exists(col => columnEquals(f.name, col))) {
-        fillCol[T](f, value)
-      } else {
-        df.col(f.name)
-      }
+      typeMatches && cols.exists(col => columnEquals(f.name, col))
+    }.map { col =>
+      (col.name, fillCol[T](col, value))
     }
-    df.select(projections : _*)
+    df.withColumns(fillColumnsInfo.map(_._1), fillColumnsInfo.map(_._2))
```

Comment:
When `df` has a duplicate column name, what is the behavior? Also, we need to add test cases to ensure the behaviors are consistent.

Reply:
When we fill the duplicate column, we'll still get `AnalysisException: Reference xx is ambiguous`. Add test cases in 03305be.